dali/operators/decoder/audio/audio_decoder_op.cc

```diff
@@ -98,95 +92,61 @@ AudioDecoderCpu::SetupImpl(std::vector<OutputDesc> &output_desc, const workspace
-
-    tmp_buf.resize(tmp_size);
-    size_t num_samples = meta.length * meta.channels;
-    size_t ret = decoders_[sample_idx]->Decode(as_raw_span(tmp_buf.data(), num_samples));
-    DALI_ENFORCE(ret == num_samples,
-                 make_string("Error decoding audio file: ", files_names_[sample_idx]));
-
-    if (should_downmix) {
-      if (should_resample) {
-        // downmix and resample
-        float *downmixed = tmp_buf.data() + meta.length * meta.channels;
-        assert(downmixed + meta.length <= tmp_buf.data() + tmp_buf.size());
-        kernels::signal::Downmix(downmixed, tmp_buf.data(), meta.length, meta.channels);
-        resampler_.Resample(audio.data, 0, audio.shape[0], output_rate,
-                            downmixed, meta.length, meta.sample_rate);
-      } else {
-        // downmix only
-        kernels::signal::Downmix(audio.data, tmp_buf.data(), meta.length, meta.channels);
-      }
-    } else if (should_resample) {
-      // multi-channel resample
-      resampler_.Resample(audio.data, 0, audio.shape[0], output_rate,
-                          tmp_buf.data(), meta.length, meta.sample_rate, meta.channels);
-
-    } else {
-      // convert or copy only - this will only happen if resampling is specified, but this
-      // recording's sampling rate and number of channels coincides with the target
-      int64_t len = std::min<int64_t>(volume(audio.shape), meta.length*meta.channels);
-      for (int64_t ofs = 0; ofs < len; ofs++) {
-        audio.data[ofs] = ConvertSatNorm<OutputType>(tmp_buf[ofs]);
-      }
-    }
-  } else {
-    assert(!should_downmix && !should_resample);
-    size_t num_samples = volume(audio.shape);
-    size_t ret = decoders_[sample_idx]->Decode(as_raw_span(audio.data, num_samples));
-    DALI_ENFORCE(ret == num_samples,
-                 make_string("Error decoding audio file: ", files_names_[sample_idx]));
-  }
+  int64_t decode_scratch_sz = 0;
+  int64_t resample_scratch_sz = 0;
+  if (should_resample || should_downmix || !std::is_same<OutputType, DecoderOutputType>::value)
+    decode_scratch_sz = meta.length * meta.channels;
+
+  // resample scratch is used to prepare a single or multiple (depending if
+  // downmixing is needed) channel float input, required by the resampling
+  // kernel
+  int64_t out_channels = should_downmix ? 1 : meta.channels;
+  if (should_resample)
+    resample_scratch_sz = meta.length * out_channels;
+
+  int64_t total_scratch_sz =
+      decode_scratch_sz * sizeof(DecoderOutputType) + resample_scratch_sz * sizeof(float);
+  scratch_[thread_idx].resize(total_scratch_sz);
+  uint8_t* scratch_mem = scratch_[thread_idx].data();
+
+  span<DecoderOutputType> decoder_scratch_mem(reinterpret_cast<DecoderOutputType *>(scratch_mem),
+                                              decode_scratch_sz);
+  span<float> resample_scratch_mem(
+        reinterpret_cast<float *>(scratch_mem + decode_scratch_sz * sizeof(DecoderOutputType)),
+        resample_scratch_sz);
```

Comment:
Please reverse the order of these - otherwise you might end up with misaligned floats which will severely degrade the performance.

Reply:
split to two different buffers